bee-identity/src/identity.rs

```diff
@@ -0,0 +1,130 @@
+    pub fn public_key(&self) -> Ed25519PublicKey {
+        self.keypair.public()
+    }
+
+    /// Returns the corresponding [`libp2p_core::PeerId`].
+    pub fn peer_id(&self) -> PeerId {
+        PeerId::from_public_key(&PublicKey::Ed25519(self.keypair.public()))
+    }
+
+    /// Returns whether this identity has been restored from a file.
+    pub fn is_restored(&self) -> bool {
+        self.restored
+    }
+
+    /// Stores this identity to a file.
+    pub fn store(&self, identity_path: impl AsRef<Path>) -> Result<(), Box<dyn std::error::Error>> {
+        pem_file::write_keypair_to_pem_file(identity_path, &self.keypair).map_err(|e| {
+            log::error!("Failed to write PEM file: {}", e);
+            e
+        })?;
+
+        Ok(())
+    }
+
+    /// Tries to restore an identity from a PEM file. If that doesn't exist or can't be restored generates a new
+    /// identity.
+    pub fn restore_or_new(
+        identity_path: impl AsRef<Path>,
+        deprecated_identity_field: Option<String>,
+    ) -> Result<Self, Box<dyn std::error::Error>> {
+        match pem_file::read_keypair_from_pem_file(&identity_path) {
+            Ok(keypair) => {
+                if deprecated_identity_field.is_some() {
+                    log::warn!(
+                        "The config file contains an `identity` field which will be ignored. You may safely delete this field to suppress this warning."
+                    );
+                }
+                Ok(Identity::from_keypair(keypair))
+            }
+            Err(pem_file::PemFileError::Read(_)) => {
+                // If we can't read from the file (which means it probably doesn't exist) we either migrate from the
+                // existing config or generate a new identity.
+                let identity = if let Some(identity_encoded) = deprecated_identity_field {
+                    log::warn!(
+                        "There is no identity file at `{}`. Migrating identity from the existing config file.",
+                        identity_path.as_ref().display(),
+                    );
+
+                    let keypair = migration::migrate_keypair(identity_encoded).map_err(|e| {
+                        log::error!("Failed to migrate keypair: {}", e);
+                        e
+                    })?;
+
+                    Identity::from_keypair(keypair)
+                } else {
+                    log::info!(
+                        "There is no identity file at `{}`. Generating a new one.",
+                        identity_path.as_ref().display()
+                    );
+
+                    Identity::new()
+                };
+
+                identity.store(identity_path)?;
+
+                Ok(identity)
+            }
+            Err(e) => {
+                log::error!("Could not extract private key from PEM file: {}", e);
+                Err(Box::new(e))
+            }
+        }
+    }
+}
+
+impl Default for Identity {
+    fn default() -> Self {
```

Comment:
I think we once talked about that and preferred to have a `generate` method instead of a Default impl

Reply:
true!

Reply:
Fixed.